### PR TITLE
Refactor hardware wallet

### LINKED ISSUE #4790: Refactor hardware wallet
### Description
After gather all HW function to the HW manager and Client and move the functionality of the HW utility must be remove from the source code. [Overview](https://miro.com/app/board/uXjVPsEwknc=/?share_link_id=99516765577).

### Acceptance Criteria
- [ ] Ledger / Server - Set up folder structure & delete old IPC logic (Trezor, Ledger, utils)
- [ ] Ledger / Server - Setup IPC main channel to talk to liskApp
- [ ] Ledger / Server - Setup IPC listen for ledger device events
- [ ] Ledger / Server - getSignedTransaction(devicePath, accountIndex, unsignedMessage)
- [ ] Ledger / Server - getIsAppOpen(devicePath, accountIndex)
- [ ] Ledger / Server - getSignedMessage(devicePath, accountIndex, unsignedMessage)
- [ ] Ledger / Server - getPubKey(devicePath, accountIndex)
- [ ] Ledger / Server - Implement queuing system to talk to the ledger
- [ ] Ledger / Server - Preserve connected ledger state on page reload and on electron start
- [ ] Ledger / Client - New Redux state: hwDevice: { currentDevice, connectedDevices }
- [ ] Ledger / Client - Toast device notification (selectDeviceModal link if +1 devices)
- [ ] Ledger / Client - Update Redux state: account: { ..., hwAccounts } - Part 1 (single account)
- [ ] Ledger / Client - Switch device from topbar HW tooltip
- [ ] Ledger / Client - Sign transaction
- [ ] Show/Hide hardware wallet icon in top bar, hide when no device are connected

### PR DESCRIPTION
**Pull request by oskarleonard**
### What was the problem?

This PR resolves #4790, #4791

### How was it solved?

- [x] Ledger / Server - Set up folder structure
- [x] Ledger -  delete old IPC logic (Trezor, Ledger, utils)
- [x] Ledger / Server - Setup IPC main channel to talk to liskApp
- [x] Ledger / Server - Setup IPC listen for ledger device events
- [x] Ledger / Server - getSignedTransaction(devicePath, accountIndex, unsignedMessage)
- [x] Ledger / Server - getSignedMessage(devicePath, accountIndex, unsignedMessage)
- [x] Ledger / Server - getPubKey(devicePath, accountIndex)
- [x] Ledger / Server - Implement queuing system to talk to the ledger
- [x] Ledger / Server - Preserve connected ledger state on page reload and on electron start
- [x] Ledger / Client - New Redux state: hwDevice: { currentDevice, connectedDevices }
- [x] Ledger / Client - Toast device notification (selectDeviceModal link if +1 devices)
- [x] Ledger / Client - Update Redux state: account: { ..., hwAccounts } - Part 1 (single account)
- [x] Ledger / Client - Switch device from topbar HW tooltip
- [x] Ledger / Client - Sign transaction
- [x] Show/Hide hardware wallet icon in top bar, hide when no device are connected

### How was it tested?

Manually

### CODE REVIEW

File: libs/ledgerHardwareWallet/ledgerLiskAppIPCChannel/clientLedgerHWCommunication/index.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    devicePath,
+    accountIndex,
+    unsignedMessage,
+  });
+
+  return signedTransaction;
+};
+
+export const getSignedMessage = async (devicePath, accountIndex, unsignedMessage) => {
+  const signedTransaction = await executeIPCCommand(GET_SIGNED_MESSAGE, {
```
Comment (ManuGowda) [inline libs/ledgerHardwareWallet/ledgerLiskAppIPCChannel/clientLedgerHWCommunication/index.js]:
```suggestion
  const signedMessage = await executeIPCCommand(GET_SIGNED_MESSAGE, {
```

File: libs/ledgerHardwareWallet/ledgerLiskAppIPCChannel/serverLedgerHWCommunication/index.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  } catch (error) {
+    await transport?.close();
+    return Promise.reject(error);
+  }
+}
+
+export async function getSignedTransaction({ devicePath, accountIndex, unsignedMessage }) {
+  let transport = TransportNodeHid;
+  try {
+    transport = await transport.open(devicePath);
```
Comment (ManuGowda) [inline libs/ledgerHardwareWallet/ledgerLiskAppIPCChannel/serverLedgerHWCommunication/index.js]:
```suggestion
  let transport;
  try {
    transport = await TransportNodeHid.open(devicePath);
```

please fix this everywhere 🙏🏼

Comment (oskarleonard) [reply libs/ledgerHardwareWallet/ledgerLiskAppIPCChannel/serverLedgerHWCommunication/index.js]: fixed

**Review by ManuGowda — DISMISSED** [review]
LGTM, tested all the functionality with hardware wallet.

File: libs/hardwareWallet/ledger/ledgerDeviceListener/useLedgerDeviceListener.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      (async () => {
+        try {
+          const connectedDevices = await getConnectedHWDevices();
+          if (connectedDevices?.length > 0) {
+            dispatch(setHWDevices(connectedDevices));
+            dispatch(setCurrentHWDevice(connectedDevices[0]));
+          }
+        } finally {
+          setHasCheckedHWDevices(true);
+        }
```
Comment (ikem-legend) [inline libs/hardwareWallet/ledger/ledgerDeviceListener/useLedgerDeviceListener.js:35]: @oskarleonard There should at least be a catch block for this

File: libs/hardwareWallet/ledger/ledgerLiskAppIPCChannel/clientLedgerHWCommunication/index.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+export const getIsInsideLiskApp = async (devicePath, accountIndex) => {
+  const pubKey = await getPubKey(devicePath, accountIndex);
+
+  return !!pubKey;
+};
+
+export const getConnectedHWDevices = async () => {
+  await sleep(1500);
+  const signedTransaction = await executeIPCCommand(GET_CONNECTED_DEVICES);
+  return signedTransaction;
```
Comment (ikem-legend) [inline libs/hardwareWallet/ledger/ledgerLiskAppIPCChannel/clientLedgerHWCommunication/index.js]: @oskarleonard I feel like the variable name doesn't reflect what it does. `connectedDevices` might be a more accurate name

Comment (ikem-legend) [inline libs/hardwareWallet/ledger/ledgerLiskAppIPCChannel/clientLedgerHWCommunication/index.js]: @oskarleonard I feel like the variable name doesn't reflect what it does. `connectedDevices` might be a more accurate name

File: src/modules/hardwareWallet/components/HardwareWalletStatus/HardwareWalletStatus.js [hunk 5]
```diff
@@ -2,29 +2,41 @@ import React from 'react';
...
 import { useTranslation } from 'react-i18next';
 import Icon from '@theme/Icon';
-import { useHWStatus } from '@hardwareWallet/hooks/useHWStatus';
+import { useSelector } from 'react-redux';
+import { selectCurrentHWDevice } from '@hardwareWallet/store/selectors/hwSelectors';
+import DialogLink from '@theme/dialog/link';
 import styles from './HardwareWalletStatus.css';
 
-const Status = ({ status }) => (
+/* const Status = ({ status }) => (
```
Comment (ikem-legend) [inline src/modules/hardwareWallet/components/HardwareWalletStatus/HardwareWalletStatus.js]: @oskarleonard If this was intentionally commented, a TODO should be added, else it should be removed

File: src/utils/usePrevious.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import React, { useEffect } from 'react';
+
+export function usePrevious(value) {
+  const ref = React.useRef();
+  useEffect(() => {
+    ref.current = value;
+  }, [value]);
```
Comment (ikem-legend) [inline src/utils/usePrevious.js:8]: @oskarleonard Possibly add a space between the return value and the previous code

**Comment by ikem-legend** [comment]
The hardware wallet icon on the top bar doesn't have the status indicator as shown in the design
Screenshot 2023-03-21 at 13 08 18

File: libs/hardwareWallet/ledger/ledgerLiskAppIPCChannel/clientLedgerHWCommunication/index.js (hunk 4 above)
Comment (oskarleonard) [reply libs/hardwareWallet/ledger/ledgerLiskAppIPCChannel/clientLedgerHWCommunication/index.js]: True, fixed!

Comment (oskarleonard) [reply libs/hardwareWallet/ledger/ledgerLiskAppIPCChannel/clientLedgerHWCommunication/index.js]: This is a duplicate of the above comment, right?

File: src/modules/hardwareWallet/components/HardwareWalletStatus/HardwareWalletStatus.js (hunk 5 above)
Comment (oskarleonard) [reply src/modules/hardwareWallet/components/HardwareWalletStatus/HardwareWalletStatus.js]: Fixed

File: src/utils/usePrevious.js (hunk 6 above)
Comment (oskarleonard) [reply src/utils/usePrevious.js:8]: Oh, yea, i actually want this in our linting enforced, space before return

File: libs/hardwareWallet/ledger/ledgerDeviceListener/useLedgerDeviceListener.js (hunk 3 above)
Comment (oskarleonard) [reply libs/hardwareWallet/ledger/ledgerDeviceListener/useLedgerDeviceListener.js:35]: Why? What do you suggest that the catch should do/be responsible for?

**Review by ManuGowda — APPROVED** [review]

**Review by ikem-legend — APPROVED** [review]
Amazing work 👏

**Pull request merged**